### PR TITLE
hcm: ensure data added during filter callbacks get propagated

### LINKED ISSUE #7579: HCM: buffered data ignored when onData never returns StopIteration
If any `encodeData` callbacks call `addEncodedData` and no StopIteration* return values are returned, the buffered data will never be read, as we never call `commonContinue` due to the `iteration_state_` check in `commonHandleAfterDataCallback`:

```
  if (status == FilterDataStatus::Continue) {
    if (iteration_state_ == IterationState::StopSingleIteration) {
      commonHandleBufferData(provided_data);
      commonContinue();
      return false;
    } else {
      ASSERT(headers_continued_);
    }
```

This becomes problematic because the following code misbehaves when there's only one `encodeData` callback:

```
  Http::FilterDataStatus encodeData(Buffer::Instance& data, bool end_stream) {
    encoder_callbacks_->addEncodedData(data, false);

    if (end_stream) {
      encoder_callbacks_->modifyEncodingBuffer([](auto& buffer) {
        // do something with the buffer
      });
      return Http::FilterDataStatus::Continue;
    }

    return Http::FilterDataStatus::StopIterationNoBuffer;
  }
};
```

Since the first call is has `end_stream = true`, we add data to buffer but since we return `Continue`, the buffer is never passed to the next filter.

### PR DESCRIPTION
**Pull request by snowp**
Currently, add*Data will write data to the filter buffer, but we only
ever read from said buffer if we ever transition from stopping ->
continuing. This means that if add*Data is called when every callback
returns Continue, the data is never propagated to the next filter.

This change adds a flag that is set whenever data is written to the
buffer, using this to flush the buffer when onData returns Continue even
if filter iteration was never stopped.

Fixes #7579

Signed-off-by: Snow Pettersen <snowp@squareup.com>

Risk Level: Medium
Testing: Added HCM UT
Docs Changes: n/a 
Release Notes: n/a
Fixes #7579

### CODE REVIEW

**Comment by dio** [comment]
@soya3129 could you help to do the first pass? Thank you.

File: source/common/http/conn_manager_impl.cc [hunk 1]
```diff
@@ -987,6 +987,7 @@ void ConnectionManagerImpl::ActiveStream::decodeData(
     recordLatestDataFilter(entry, state_.latest_data_decoding_filter_, decoder_filters_);
 
     state_.filter_call_state_ |= FilterCallState::DecodeData;
+    (*entry)->inline_buffered_data_ = false;
```
Comment (soya3129) [inline source/common/http/conn_manager_impl.cc]: Why we don't need (*entry)->inline_buffered_data_ = false; in ActiveStream::encodeData()?

File: source/common/http/conn_manager_impl.cc [hunk 2]
```diff
@@ -1741,7 +1744,6 @@ void ConnectionManagerImpl::ActiveStreamFilterBase::commonContinue() {
 
   ENVOY_STREAM_LOG(trace, "continuing filter chain: filter={}", parent_,
                    static_cast<const void*>(this));
-  ASSERT(!canIterate());
```
Comment (soya3129) [inline source/common/http/conn_manager_impl.cc]: I really like this solution! It's simple and it works! The only concern I have is the removal of ASSERT here and below. Is it possible to keep the assert? For example: ASSERT(!canIterate() || inline_buffered_data_)?

Comment (snowp) [reply source/common/http/conn_manager_impl.cc]: yep that makes sense. I was ripping out all the asserts that failed, makes sense to make it more exact instead

File: source/common/http/conn_manager_impl.cc (hunk 1 above)
Comment (snowp) [reply source/common/http/conn_manager_impl.cc]: we do :) good catch

**Pull request closed without merging**